### ISSUE TITLE
Fire-assisted skill attacks fail when more than two zero-sided dice are selected

### ISSUE THREAD

**Issue by cgolubi1** [buttonmen-dev/buttonmen#2689]
As noted in #2389 this attack fails:

![die_values_do_not_sum](https://user-images.githubusercontent.com/2898207/103711826-42365100-4f86-11eb-8182-60805a8c4dd6.png)

If i select only two of the zero-sided dice, it succeeds (it doesn't appear to matter which two; i tried several ways).

**Comment by blackshadowshade** [comment]
I'm not sure that this is a problem caused by #2353, since that doesn't change any of the contents of BMAttackSkill->is_assisted_attack_valid(), which is the variant used when Fire dice are available.

It's certainly going to need some further digging.

**Comment by cgolubi1** [comment]
I see what you're saying, but the fact that it succeeds if i unselect down to two `(0)` dice makes me really suspicious.

Anyway, i can't trivially give you a responderTest after all, of course, because we haven't merged custom buttons yet.

**Comment by cgolubi1** [comment]
Here's the raw test, but it won't help you:

```
    /**
     * @depends responder00Test::test_request_savePlayerInfo
     * @group fulltest_deps
     */
    public function test_interface_game_00068() {

        // responder003 is the POV player, so if you need to fake
        // login as a different player e.g. to submit an attack, always
        // return to responder003 as soon as you've done so
        $this->game_number = 68;
        $_SESSION = $this->mock_test_user_login('responder003');


        $gameId = $this->verify_api_createGame(
            array('bm_rand' => array(7, 0, 2, 12, 0, 6, 10, 0, 7, 7, 0, 3), 'bm_skill_rand' => array()),
            'responder003', 'responder004', 'CustomBM', 'CustomBM', 3
        );

        $expData = $this->squash_game_data_timestamps(json_decode('{"previousGameId": null, "gameChatEditable": false, "gameId": 17, "description": "", "creatorDataArray": {"creatorName": "responder003", "creatorId": 3}, "roundNumber": 1, "timestamp": 1609820809, "gameChatLog": [], "currentPlayerIdx": 0, "dieBackgroundType": "realistic", "gameState": "SPECIFY_DICE", "gameActionLogCount": 1, "playerWithInitiativeIdx": null, "activePlayerIdx": null, "gameSkillsInfo": {"Fire": {"code": "F", "interacts": [], "description": "Fire Dice cannot make Power Attacks. Instead, they can assist other Dice in making Skill and Power Attacks. Before making a Skill or Power Attack, you may increase the value showing on any of the attacking dice, and decrease the values showing on one or more of your Fire Dice by the same amount. For example, if you wish to increase the value of an attacking die by 5 points, you can take 5 points away from one or more of your Fire Dice. Turn the Fire Dice to show the adjusted values, and then make the attack as normal. Dice can never be increased or decreased outside their normal range, i.e., a 10-sided die can never show a number lower than 1 or higher than 10. Also, Fire Dice cannot assist other dice in making attacks other than normal Skill and Power Attacks."}, "TimeAndSpace": {"code": "^", "interacts": [], "description": "If a Time and Space Die participates in an attack and rerolls an odd number, then the player will take another turn. If multiple Time and Space dice are rerolled and show odd, only one extra turn is given per reroll."}, "Stealth": {"code": "d", "interacts": [], "description": "These dice cannot perform any type of attack other than Multi-die Skill Attacks, meaning two or more dice participating in a Skill Attack. In addition, Stealth Dice cannot be captured by any attack other than a Multi-die Skill Attack."}, "Poison": {"code": "p", "interacts": [], "description": "These dice are worth negative points. If you keep a Poison Die of your own at the end of a round, subtract its full value from your score. If you capture a Poison Die from someone else, subtract half its value from your score."}}, "gameActionLog": [{"timestamp": 1609820809, "message": "Game created by responder003", "player": ""}], "maxWins": 3, "playerDataArray": [{"prevOptValueArray": [], "optRequestArray": [], "playerColor": "#dd99dd", "roundScore": null, "waitingOnAction": false, "playerId": 3, "outOfPlayDieArray": [], "button": {"artFilename": "BMdefaultRound.png", "recipe": "F(14) (0) (7) F(14) (0) (7) F(14) (0) (7) F(14) (0) (7)", "name": "CustomBM", "originalRecipe": "F(14) (0) (7) F(14) (0) (7) F(14) (0) (7) F(14) (0) (7)"}, "capturedDieArray": [], "isChatPrivate": false, "isOnVacation": false, "activeDieArray": [{"description": "Fire 14-sided die", "skills": ["Fire"], "recipe": "F(14)", "value": null, "properties": [], "sides": 14}, {"description": "0-sided die", "skills": [], "recipe": "(0)", "value": null, "properties": [], "sides": 0}, {"description": "7-sided die", "skills": [], "recipe": "(7)", "value": null, "properties": [], "sides": 7}, {"description": "Fire 14-sided die", "skills": ["Fire"], "recipe": "F(14)", "value": null, "properties": [], "sides": 14}, {"description": "0-sided die", "skills": [], "recipe": "(0)", "value": null, "properties": [], "sides": 0}, {"description": "7-sided die", "skills": [], "recipe": "(7)", "value": null, "properties": [], "sides": 7}, {"description": "Fire 14-sided die", "skills": ["Fire"], "recipe": "F(14)", "value": null, "properties": [], "sides": 14}, {"description": "0-sided die", "skills": [], "recipe": "(0)", "value": null, "properties": [], "sides": 0}, {"description": "7-sided die", "skills": [], "recipe": "(7)", "value": null, "properties": [], "sides": 7}, {"description": "Fire 14-sided die", "skills": ["Fire"], "recipe": "F(14)", "value": null, "properties": [], "sides": 14}, {"description": "0-sided die", "skills": [], "recipe": "(0)", "value": null, "properties": [], "sides": 0}, {"description": "7-sided die", "skills": [], "recipe": "(7)", "value": null, "properties": [], "sides": 7}], "swingRequestArray": [], "hasDismissedGame": false, "gameScoreArray": {"L": 0, "W": 0, "D": 0}, "sideScore": null, "canStillWin": null, "turboSizeArray": [], "playerName": "responder003", "prevSwingValueArray": [], "lastActionTime": 1609820809}, {"prevOptValueArray": [], "optRequestArray": [], "playerColor": "#ddffdd", "roundScore": null, "waitingOnAction": true, "playerId": 5, "outOfPlayDieArray": [], "button": {"artFilename": "BMdefaultRound.png", "recipe": "pd^(R)", "name": "CustomBM", "originalRecipe": "pd^(R)"}, "capturedDieArray": [], "isChatPrivate": false, "isOnVacation": false, "activeDieArray": [{"description": "Poison Stealth TimeAndSpace R Swing Die", "skills": ["Poison", "Stealth", "TimeAndSpace"], "recipe": "pd^(R)", "value": null, "properties": [], "sides": null}], "swingRequestArray": {"R": [2, 16]}, "hasDismissedGame": false, "gameScoreArray": {"L": 0, "W": 0, "D": 0}, "sideScore": null, "canStillWin": null, "turboSizeArray": [], "playerName": "responder004", "prevSwingValueArray": [], "lastActionTime": 1609820809}], "gameChatLogCount": 0, "validAttackTypeArray": []}', TRUE));
        $expData['gameId'] = $gameId;
        $expData['playerDataArray'][0]['playerId'] = $this->user_ids['responder003'];
        $expData['playerDataArray'][1]['playerId'] = $this->user_ids['responder004'];

        $retval = $this->verify_api_loadGameData($expData, $gameId, 10);

        $_SESSION = $this->mock_test_user_login('responder004');
        $this->verify_api_submitDieValues(
            array(8),
            $gameId, 1, array('R' => 16), NULL);

        $_SESSION = $this->mock_test_user_login('responder003');
        $expData['activePlayerIdx'] = 0;
        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => 'responder004', 'message' => 'responder004 set swing values: R=16'));
        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => '', 'message' => 'responder003 won initiative for round 1. Initial die values: responder003 rolled [F(14):7, (0):0, (7):2, F(14):12, (0):0, (7):6, F(14):10, (0):0, (7):7, F(14):7, (0):0, (7):3], responder004 rolled [pd^(R=16):8].'));
        $expData['gameActionLogCount'] = 3;
        $expData['gameState'] = "START_TURN";
        $expData['playerDataArray'][0]['activeDieArray'][0]['value'] = 7;
        $expData['playerDataArray'][0]['activeDieArray'][1]['value'] = 0;
        $expData['playerDataArray'][0]['activeDieArray'][2]['value'] = 2;
        $expData['playerDataArray'][0]['activeDieArray'][3]['value'] = 12;
        $expData['playerDataArray'][0]['activeDieArray'][4]['value'] = 0;
        $expData['playerDataArray'][0]['activeDieArray'][5]['value'] = 6;
        $expData['playerDataArray'][0]['activeDieArray'][6]['value'] = 10;
        $expData['playerDataArray'][0]['activeDieArray'][7]['value'] = 0;
        $expData['playerDataArray'][0]['activeDieArray'][8]['value'] = 7;
        $expData['playerDataArray'][0]['activeDieArray'][9]['value'] = 7;
        $expData['playerDataArray'][0]['activeDieArray'][10]['value'] = 0;
        $expData['playerDataArray'][0]['activeDieArray'][11]['value'] = 3;
        $expData['playerDataArray'][0]['canStillWin'] = true;
        $expData['playerDataArray'][0]['roundScore'] = 42;
        $expData['playerDataArray'][0]['sideScore'] = 38.7;
        $expData['playerDataArray'][0]['waitingOnAction'] = true;
        $expData['playerDataArray'][1]['activeDieArray'][0]['description'] = "Poison Stealth TimeAndSpace R Swing Die (with 16 sides)";
        $expData['playerDataArray'][1]['activeDieArray'][0]['sides'] = 16;
        $expData['playerDataArray'][1]['activeDieArray'][0]['value'] = 8;
        $expData['playerDataArray'][1]['canStillWin'] = true;
        $expData['playerDataArray'][1]['roundScore'] = -16;
        $expData['playerDataArray'][1]['sideScore'] = -38.7;
        $expData['playerDataArray'][1]['swingRequestArray'] = array();
        $expData['playerDataArray'][1]['waitingOnAction'] = false;
        $expData['playerWithInitiativeIdx'] = 0;
        $expData['validAttackTypeArray'] = array("Skill");

        $retval = $this->verify_api_loadGameData($expData, $gameId, 10);

        // BUG encountered while logging game: API submit_turn(17, 0, 1, {'playerIdx_0_dieIdx_8': 'false', 'playerIdx_0_dieIdx_9': 'false', 'playerIdx_0_dieIdx_6': 'false', 'playerIdx_0_dieIdx_7': 'true', 'playerIdx_0_dieIdx_4': 'true', 'playerIdx_0_dieIdx_5': 'false', 'playerIdx_0_dieIdx_2': 'true', 'playerIdx_0_dieIdx_3': 'false', 'playerIdx_0_dieIdx_0': 'false', 'playerIdx_0_dieIdx_1': 'true', 'playerIdx_0_dieIdx_10': 'true', 'playerIdx_0_dieIdx_11': 'true', 'playerIdx_1_dieIdx_0': 'true'}, Skill, 1, 1609820809, {}) unexpectedly failed: Attacking die values do not sum up to target die value.
    }
```

**Comment by blackshadowshade** [comment]
I'm thinking it may be a subsequent pull request that references the same code. When I go digging, I'll do a git blame on any suspicious calls and track it back.

**Comment by blackshadowshade** [comment]
It should be testable at the BMAttackSkill level, so I'll try that first.

**Comment by blackshadowshade** [comment]
Right, I've found the issue. It is indeed related to #2353, and caused exactly because I *didn't* alter the contents of BMAttackSkill->is_assisted_attack_valid() in the same way as BMAttackSkill->is_direct_attack_valid().

I've incorporated the fix into #2389.

**Comment by jl8e** [comment]
Since we've since changed the behavior of the trip-radioactive interaction, does it make sense to keep the zero-stripping behavior at all?

**Comment by blackshadowshade** [comment]
I don't remember changing the behaviour of trip radioactive. Remind me where?

**Comment by blackshadowshade** [comment]
Also, yes I think it makes sense to keep the zero-stripping behaviour, it has no impact on the logic for determining if a skill attack is valid since all these zeros were are stripping have no skills.

**Comment by jl8e** [comment]
> I don't remember changing the behaviour of trip radioactive. Remind me where?

#2436

**Comment by blackshadowshade** [comment]
Ah, thank you, Julian.

I think that as we progress CustomBM, we're going to be removing some of the restrictions that we've imposed on the current pull request, and so the problem of too many zeros may actually be one that will potentially continue to exist.

I'm going to have to do a more generalised restriction of the size of the hit table to address the issue of potentially large numbers of sides, which I'll get onto after we've got CustomBM accepted with the current restrictions. I admit that *this* will almost certainly have the potential to create weird bugs since we've already seen it especially with negative konstant and fire, but I think these are bugs we need to tackle and not avoid.
